src/decoder/lattice-faster-decoder-combine-bucketqueue.cc

```diff
@@ -0,0 +1,1085 @@
+// decoder/lattice-faster-decoder-combine.cc
+
+// Copyright 2009-2012  Microsoft Corporation  Mirko Hannemann
+//           2013-2019  Johns Hopkins University (Author: Daniel Povey)
+//                2014  Guoguo Chen
+//                2018  Zhehuai Chen
+//                2019  Hang Lyu
+
+// See ../../COPYING for clarification regarding multiple authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#include "decoder/lattice-faster-decoder-combine.h"
+#include "lat/lattice-functions.h"
+
+namespace kaldi {
+
+template<typename Token>
+BucketQueue<Token>::BucketQueue(BaseFloat cost_scale) :
+    cost_scale_(cost_scale) {
+  // NOTE: we reserve plenty of elements to avoid expensive reallocations
+  // later on. Normally, the size is a little bigger than (adaptive_beam +
+  // 15) * cost_scale.
+  int32 bucket_size = 100;
+  buckets_.resize(bucket_size);
+  bucket_offset_ = 15 * cost_scale_;
+  first_nonempty_bucket_index_ = bucket_size - 1;
+  first_nonempty_bucket_ = &buckets_[first_nonempty_bucket_index_];
+}
+
+template<typename Token>
+void BucketQueue<Token>::Push(Token *tok) {
+  size_t bucket_index = std::floor(tok->tot_cost * cost_scale_) +
+                        bucket_offset_;
+  if (bucket_index >= buckets_.size()) {
+    int32 margin = 10;  // a margin which is used to reduce re-allocate
+                        // space frequently
+    if (static_cast<int32>(bucket_index) > 0) {
+      buckets_.resize(bucket_index + margin);
+      first_nonempty_bucket_ = &buckets_[first_nonempty_bucket_index_];
+    } else {  // less than 0
+      int32 increase_size = - static_cast<int32>(bucket_index) + margin;
+      buckets_.resize(buckets_.size() + increase_size);
+      // translation
+      for (size_t i = buckets_.size() - 1; i >= increase_size; i--) {
+        buckets_[i].swap(buckets_[i - increase_size]);
+      }
+      bucket_offset_ = bucket_offset_ + increase_size * cost_scale_;
```

Comment:
This line doesn't look right to me, there should be no `* cost_scale_`; bucket_offset_ is a number of buckets.
It might affect decoding results (although it would only impact frames on which the bucket list was resize).